focused_proof.ml

```diff
@@ -402,77 +397,77 @@ let rec unfocalize_proof theta = function
-                    contract exponentials_together [] (gamma @ [Tensor (e1,e2)]) theta
+                    contract exponentials_together [] (gamma @ [Tensor (e1,e2)]) theta_elements
                 | _ -> raise (Failure "sync tensor expected") end
             | Plus_left_intro -> begin match focused_sequent with
-                | Sync (_, gamma, Plus (e1, e2)) ->
-                   Plus_left_proof (map_wn theta @ gamma, e1, e2, [], unfocalize_proof theta (List.hd focused_premises))
-               | _ -> raise (Failure "async with plus expected") end
-          | Plus_right_intro -> begin match focused_sequent with
-               | Sync (_, gamma, Plus (e1, e2)) ->
-                   Plus_right_proof (map_wn theta @ gamma, e1, e2, [], unfocalize_proof theta (List.hd focused_premises))
-               | _ -> raise (Failure "async with plus expected") end
-          | Ofcourse_intro -> begin match focused_sequent with
-               | Sync (_, [], Ofcourse e) ->
-                   Promotion_proof (theta, e, [], unfocalize_proof theta (List.hd focused_premises))
+                | Sync (theta, gamma, Plus (e1, e2)) ->
+                    Plus_left_proof (map_wn (Set_formula.elements theta) @ gamma, e1, e2, [], unfocus_proof (List.hd focused_premises))
+                | _ -> raise (Failure "async with plus expected") end
+            | Plus_right_intro -> begin match focused_sequent with
+                | Sync (theta, gamma, Plus (e1, e2)) ->
+                    Plus_right_proof (map_wn (Set_formula.elements theta) @ gamma, e1, e2, [], unfocus_proof (List.hd focused_premises))
+                | _ -> raise (Failure "async with plus expected") end
+            | Ofcourse_intro -> begin match focused_sequent with
+                | Sync (theta, [], Ofcourse e) ->
+                   Promotion_proof (Set_formula.elements theta, e, [], unfocus_proof (List.hd focused_premises))
                 | _ -> raise (Failure "sync with ofcourse expected") end
             | Whynot_intro -> begin match focused_sequent with
-                | Async (_, gamma, (Whynot e) :: tail) ->
-                    let premise = unfocalize_proof (e::theta) (List.hd focused_premises) in
-                    let new_sequent = map_wn (e::theta) @ gamma @ tail in
-                    let n_theta = List.length theta in
-                    let n_gamma = List.length gamma in
-                    let permutation = List.init n_theta (fun k -> 1 + k)
-                        @ List.init n_gamma (fun k -> 1 + n_theta + k)
-                        @ [0]
-                        @ List.init (List.length tail) (fun k -> 1 + n_theta + n_gamma + k) in
-                    Exchange_proof (new_sequent, permutation, premise)
+                | Async (theta, gamma, (Whynot e) :: tail) ->
+                    if Set_formula.mem e theta then
+                      Weakening_proof (map_wn (Set_formula.elements theta) @ gamma, e, tail, unfocus_proof (List.hd focused_premises))
```

Comment:
What does it bring? This case will be handled by the "weakening simplification" at the end, won't it?

Reply:
understood, it is to avoid doublon in theta.

Reply:
It `theta` is not given as an argument anymore, this allows to preserve the invariant that `unfocus_proof` is a proof with conclusion `map_wn (Set_formula.elements theta) @ gamma @ lambda` if the focused proofs has conclusion `Async (theta, gamma, lambda)`.